src/server/races/default-race.js

```diff
@@ -0,0 +1,279 @@
+const gameConstants = require("../../game-constants");
+const config = require("../config");
+const Marble = require("./marble");
+const RaceEntry = require("./race-entry");
+const db = require("../database/manager");
+const physicsWorld = require("../../physics/world");
+const skins = require("../skins");
+const permissions = require("../chat/permissions");
+const utility = require("../../utility");
+
+// Object containing all race data to be recorded in the database
+function Round(levelId) {
+	this.start = null;
+	this.end = null;
+	this.timeBest = null;
+	this.levelId = levelId;
+	this.pointsAwarded = 0;
+	this.playersEntered = 0;
+	this.playersFinished = 0;
+	this.playersNotFinished = 0;
+	this.marblesEntered = 0;
+	this.marblesFinished = 0;
+	this.marblesNotFinished = 0;
+}
+
+class DefaultRace {
+	constructor(game, levelId) {
+		this.game = game;
+		this.raceEntries = [];
+		this.pointContributors = 0;
+		this.marbleEntries = [];
+		this.marbleLimitReached = false;
+		this.marblesFinished = 0;
+
+		this.round = new Round(levelId);
+
+		// Contains a set of functions that returns a valid value if the passed attribute string can be parsed correctly
+		this.marbleAttributeList = {};
+		this.defaultMarbleAttributeList = {}; // Similar to above, but sets defaults for attributes that were not set/provided by the user
+
+		// Define default marble attributes
+		this.marbleAttributeList.color = function(attribute) {
+			const colorRegEx = /#(?:[0-9a-fA-F]{3}){1,2}$/g;
+			let match = attribute.match(colorRegEx);
+			return match === null ? undefined : match[0];
+		};
+		this.defaultMarbleAttributeList.color = function() {
+			return utility.randomHexColor();
+		};
+
+		this.marbleAttributeList.skinId = function(attribute, userId) {
+			let skinId = skins.idList[attribute];
+			if (typeof skinId === "undefined")
+				return undefined;
```

Comment:
Is it better, or is it preference?
Personally, I don't like conditionals where a lot of different/unrelated evaluations are done at once, because it makes it harder to read and understand what is being checked (and in what order the checks need to happen, if applicable). Keeping this `if` as an early-out and the one below as purely a permission check seems like more readable code to me.

Also, combining these adds an `||` operator and requires the permission check to be wrapped in parentheses, making it even less readable in my opinion.

Reply:
Fair points, we can keep it as it is 👍